feat: Add the creation of `Microsoft.KeyVault/vaults/secrets` in Sql Server Module - `avm/res/sql/server`

jianingwang123 opened the pull request:
## Description

<!--
>Thank you for your contribution !
> Please include a summary of the change and which issue is fixed.
> Please also include the context.
> List any dependencies that are required for this change.

Fixes #123
Fixes #456
Closes #123
Closes #456
-->
Fixes #2608

Set `azure/login@v2` to use federated credential. (Due to some security policies, we are currently unable to use client secret for azure login. After completing all changes to this PR, we will restore the login method. This is currently only for testing the github pipeline.)

## Pipeline Reference

<!-- Insert your Pipeline Status Badge below -->

| Pipeline |
| -------- |
|    [![avm.res.sql.server](https://github.com/jianingwang123/bicep-registry-modules/actions/workflows/avm.res.sql.server.yml/badge.svg?branch=fix%2F2608)](https://github.com/jianingwang123/bicep-registry-modules/actions/workflows/avm.res.sql.server.yml)      |

## Type of Change

<!-- Use the checkboxes [x] on the options that are relevant. -->

- [ ] Update to CI Environment or utilities (Non-module affecting changes)
- [x] Azure Verified Module updates:
  - [ ] Bugfix containing backwards-compatible bug fixes, and I have NOT bumped the MAJOR or MINOR version in `version.json`:
    - [ ] Someone has opened a bug report issue, and I have included "Closes #{bug_report_issue_number}" in the PR description.
    - [ ] The bug was found by the module author, and no one has opened an issue to report it yet.
  - [ ] Feature update backwards compatible feature updates, and I have bumped the MINOR version in `version.json`.
  - [ ] Breaking changes and I have bumped the MAJOR version in `version.json`.
  - [ ] Update to documentation

## Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [x] I have run `Set-AVMModule` locally to generate the supporting module files.
- [x] My corresponding pipelines / checks run clean and green without any errors or warnings

<!--  Please keep up to date with the contribution guide at https://aka.ms/avm/contribute/bicep -->

@jongio - for notification.

jtracey93 commented:
@AlexanderSehr, @eriqua this PR changes the workflow for the module towards OIDC, but i am aware @eriqua has a PR due to sort this across all modules. Can you advise what we do here.

Also @jianingwang123 & @jongio as per my comment [here](https://github.com/Azure/bicep-registry-modules/issues/2608#issuecomment-2226344022) is there any chance you or the team want to take ownership of this module?

jongio commented:
> Also @jianingwang123 & @jongio as per my comment [here](https://github.com/Azure/bicep-registry-modules/issues/2608#issuecomment-2226344022) is there any chance you or the team want to take ownership of this module?

I'm pinging some people to see if we can find an owner.

AlexanderSehr commented on .github/actions/templates/avm-validateModuleDeployment/action.yml:
```diff
@@ -62,7 +62,9 @@ runs:
       uses: azure/login@v2
       with:
-        creds: ${{ env.AZURE_CREDENTIALS }}
+        client-id: ${{ env.AZURE_CLIENT_ID }}
```
Remove

AlexanderSehr commented on .github/workflows/avm.template.module.yml:
```diff
@@ -25,7 +25,10 @@ env:
   ARM_TENANT_ID: "${{ secrets.ARM_TENANT_ID }}"
   TOKEN_NAMEPREFIX: "${{ secrets.TOKEN_NAMEPREFIX }}"
-
+  AZURE_CLIENT_ID: "${{ secrets.AZURE_CLIENT_ID }}"
```
Remove

jianingwang123 replied:
> Remove OIDC changes

This is for testing purposes only. After all changes are complete, we will revert the changes to these two files.

AlexanderSehr replied:
Sounds fair. Thanks @jianingwang123

AlexanderSehr commented:
> Set `azure/login@v2` to use federated credential

OIDC 

> Member

Correct. OIDC should not be part of a module PR given that it's a CI topic

AlexanderSehr commented on avm/res/sql/server/main.bicep:
```diff
@@ -485,6 +488,41 @@ module server_audit_settings 'audit-settings/main.bicep' = if (!empty(auditSetti
   }
 }
 
+module server_secrets_export 'modules/key-vault-export.bicep' = if (!empty(secretsExportConfiguration)) {
```
This change should be put on hold until https://github.com/Azure/bicep-registry-modules/issues/1934 is resolved and the spec defined. @jtracey93, added it for the upcoming discussion call. Shouldn't take long to go through the suggestions, agree to one and define the spec. There are also a few other PRs that were created similar to this one in the past:
- #1920 
- #1932 
- #1950

We have to get this spec out there asap.

AlexanderSehr commented on avm/res/sql/server/main.bicep:
```diff
@@ -652,3 +693,26 @@ type auditSettingsType = {
+  @description('Optional. The name for secret to create.')
+  appUserPasswordSecretName: string?
+
+  @description('Optional. The value for secret to create.')
+  appUserPasswordSecretValue: string?
```
Should likely not be part of the spec. Instead it would make more sense to me to add the secret outside of the module given that it's not connected to any output of the SQL-Server, but instead an input

zedy-wj replied:
After understanding, we initially thought that there would be some built-in methods where the `Sql Server` is created, such as: `ListConnectionString()`, so that we can set the `KeyVault Secrets` more conveniently than splicing it.

But the current approach is no different from our external splicing, so I agree that it is more appropriate to store these `Secrets` outside the module.😃

AlexanderSehr replied:
Hey @zedy-wj that would indeed be the idea. I already commented on the linked issue [here](https://github.com/Azure/bicep-registry-modules/issues/2608#issuecomment-2255084879).

It makes sense to set service-specific secrets like connections strings etc. as you cannot access them from outside the module without an `existing` reference.
But adding also custom secrets to the mix (i.e., those which do not come from the service), seemingly goes a bit too far of topic.

zedy-wj replied:
Hey @AlexanderSehr - Based on the current file changes, it looks like `appUserPasswordSecretValue` is part of the `connection string`. Is there a better way to get the `connection string` if we don't pass this value externally? It seems to me that currently setting `appUserPasswordSecretValue` as a `KeyVault Secret` is just for convenience (Of course, it seems more appropriate to do this outside the module at present) since it's also required in the `connection string`.😄

AlexanderSehr replied:
Fair point. I think the below as a value defined in the module that is stored in a key vault makes sense, as it references information that is only available from inside the module, and that it also neds the appUserName + password seems like a service-specific special case to me.

```bicep
{
  name: secretsExportConfiguration.?sqlAzureConnectionStringSercretName
  value: 'Server=${server.properties.fullyQualifiedDomainName}; Database=${!empty(databases) ? databases[0].name : ''}; User=${secretsExportConfiguration.?appUserName}; Password=${secretsExportConfiguration.?appUserPasswordSecretValue}'
}
```
the wordking of the `appUserName` & `appUserPasswordSecretValue` should potentially be altered a bit to make it more obvious that they belong to the `sqlAzureConnectionString` secret, but aside from that it seems to be the best option.

The only property I'd argue against is only this configuration:
```bicep
{
  name: secretsExportConfiguration.?appUserPasswordSecretName
  value: secretsExportConfiguration.?appUserPasswordSecretValue
}
``` 
because it does not reference anything from the service but could just as well be set outside the module.

zedy-wj replied:
The latest update has been pushed, and it seems that using `administratorLogin` and `administratorLoginPassword` still works well. In this way, some inputs will not be necessary.

AlexanderSehr commented on avm/res/sql/server/main.bicep:
```diff
@@ -652,3 +689,17 @@ type auditSettingsType = {
+type secretsExportConfigurationType = {
+  @description('Required. The resource ID of the key vault where to store the secrets of this module.')
+  keyVaultResourceId: string
+
+  @description('Optional. Default to the resource group where this account is. The resource group name where the key vault is.')
+  resourceGroupName: string?
```
However, in general, maybe we should put any review on hold until the spec is actually finalized. To ask you to make changes towards something that may not be final seem a bit ... wrong. I hope, by Thursday, this will change.

bryansan-msft commented:
Im not sql server module owner anymore

eriqua commented:
Hi @jianingwang123, after a few iterations, the bicep secrets export spec has finally been finalized. You can find the schema published [here](https://azure.github.io/Azure-Verified-Modules/specs/bicep/interfaces/).

Would you be available to review it and leverage it to drive this PR to closure?
Please reach out if any doubts.

ReneHezser commented:
@jianingwang123 
We have added examples for Bicep parameter files to the Readme. This has been applied to all published modules but needs to be done for PRs as well. Can you please run the Set-AVMModule utility as detailed [here](https://azure.github.io/Azure-Verified-Modules/contributing/bicep/bicep-contribution-flow/generate-bicep-module-files/). It is required for the validation pipeline to succeed and the contribution to be published.

Please reach out if any support is needed.

jianingwang123 commented:
> Hi @jianingwang123, after a few iterations, the bicep secrets export spec has finally been finalized. You can find the schema published [here](https://azure.github.io/Azure-Verified-Modules/specs/bicep/interfaces/).
> 
> Would you be available to review it and leverage it to drive this PR to closure? Please reach out if any doubts. Thanks!

@AlexanderSehr and @eriqua, I have updated based on the bicep secrets export spec. Please help review.

jianingwang123 commented:
> @jianingwang123 We have added examples for Bicep parameter files to the Readme. This has been applied to all published modules but needs to be done for PRs as well. Can you please update your branch and run the Set-AVMModule utility as detailed [here](https://azure.github.io/Azure-Verified-Modules/contributing/bicep/bicep-contribution-flow/generate-bicep-module-files/). It is required for the validation pipeline to succeed and the contribution to be published.
> 
> Please reach out if any support is needed.

@ReneHezser Thank you for your guidance. I have updated branch and re-ran the Set-AVMModule utility.

AlexanderSehr commented:
Hey @peterbud and @emichellecarter, please give this PR a spin when you get the chance :)

peterbud reviewed (approved):
Thanks!
